### DISCUSSION TITLE
Adding guards causes inexplicable behaviour changes elsewhere

### DISCUSSION THREAD

**Discussion by AstridHornBrorholt** [UPPAALModelChecker/UPPAAL-Meta#166] [Q&A] [answered]
Version: `UPPAAL 4.1.20-stratego-11-beta3 (rev. 6F99EE8A8C5FB4A1), November 2022`. Also tested on stratego 10-beta-10 and stratego 10-release. My system is `Ubuntu 22.04.1 LTS 64-bit`. 

I suspect the issue is with the symbolic simulation. Either I have been allowed by the UI to do something inadvisable, or it is due to a bug.

I am using a stripped down version of the  [Cruise Control](https://uppaal.org/casestudies/stratego/#safe-and-optimal-cruise-control)
 example, where I have removed the updateDiscrete() function and associated integer values. This leaves only the hybrid clock simulation of the system, and I have updated the guards to reflect this. (as they previously used the integer values)
My changes appear to work fine and can be seen in [CC_OK.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/9970998/CC_OK.zip)

An example trace can be seen here: 

![Untitled2](https://user-images.githubusercontent.com/11016262/200833791-d2a720f5-eec3-4b1b-93e6-9f06b741cc23.png)

However, when I add two more edges with certain guards, suddenly two unrelated behavior changes occur in the query results. The changed model is the following: [CC_bad.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/9971130/CC_bad.zip)

The guards are `rVelocityEgo >= maxVelocityEgo` and `rVelocityEgo <= minVelocityEgo` which I have given their own edges to lead to the neutral action. Separate edges are used because it errors if I use `||` with hybrid clock guards. 

Before:
![image](https://user-images.githubusercontent.com/11016262/200832745-d9c9b2da-6cfd-4b35-a668-a5aeb41edffa.png)

After: 
![image](https://user-images.githubusercontent.com/11016262/200833521-7c36fcdd-5041-4f6f-9bb7-ab59c51c0467.png)

And the unexpected behavior is:

- `rVelocityEgo` will suddenly drop to the value of `minVelocityEgo` in one time-step.
- The model will apparently stop picking negative accelerations for Ego. 

As can be seen in this query trace:

![Untitled](https://user-images.githubusercontent.com/11016262/200831208-7903997e-ea8a-41ae-aaf4-6b2923d79dfd.png)

Note the sudden drop in the yellow `rVelocityEgo` line at 6 seconds. 

Adding the guards is just a step towards applying some other restrictions on Ego, and I cannot see how they should change the model's behavior except perhaps change the distribution of random outcomes slightly.

**Comment by mikucionisaau** [comment, marked as answer]
1) The issue is with concrete simulator (not symbolic).
2) You may want to enable `JFreeChart` in the `Edit`/`Preferences`, which allows zoom-in and zoom-out in the plots.
2) I can reproduce the sudden drop in `rVelocityEgo` also in `CC_OK` model, so `CC_bad` is not special in that regard:
![image](https://user-images.githubusercontent.com/4868894/201092278-61d9ef55-33d7-4285-aadb-681cf3d7a38b.png)

I cannot find an explanation for this drop by -8 in no time -- it must be a bug. There is also another instant drop at 3.0, cannot explain that either.

I attach a minimal counter-example [CC_bug.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/9981096/CC_bug.zip) which demonstrates the sudden jump in hybrid clock value which should not happen. 
It seems that the trace-generator did not check for committed/urgent location and applied the delay needed to satisfy the guard, which is certainly a bug.

**Reply by mikucionisaau** [reply]
Fixed in Stratego-11 rc1

**Reply by AstridHornBrorholt** [reply]
I have tried it again in Stratego-11 rc1 and it works as expected!
